### PR TITLE
[SPARK-49082][SQL] Widening type promotions in `AvroDeserializer`

### PR DESCRIPTION
**Pull request by wayneguow**
### What changes were proposed in this pull request?

This PR aims to widen type promotions in `AvroDeserializer`. Supported as following(Avro Type -> Spark Type):

- Int -> Long ;
- Int -> Double ;
- Float -> Double;

### Why are the changes needed?

Similar to PR #44368 for `Parquet` reader, we'd better to enable type promotion/widening for `Avro` deserializer.

### Does this PR introduce _any_ user-facing change?

Yes, but more convenient for users.

### How was this patch tested?

Pass GA and add a new test case.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by wayneguow** [comment]
cc @cloud-fan  and @LuciferYang

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala
Comment (cloud-fan) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:942]:
```suggestion
      val floatPath = s"$tempPath/float_data"
```

**Review by cloud-fan — APPROVED** [review]

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala
Comment (cloud-fan) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:943]: can we also test +/- infinity and nan?

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala
Comment (wayneguow) [inline connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:204]: @cloud-fan  I made a change here. 

When it comes to converting maximum or minimum values ​​from float to double, if using the implicit toDouble directly, the result will not meet expectations. For example, the difference in the result is actually a very large number.

image

And I think this problem is also faced in the processing of parquet.
https://github.com/apache/spark/blob/d431d4034219f2c84c105e1894eebe03745bf105/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala#L330

Comment (wayneguow) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:204]: At the same time, I think the logic of converting float to double(for float, `case x: NumericType` is matched, and then `toDouble` is called directly) in our cast expression is also worth discussing the implementation, which is similar to this problem.

https://github.com/apache/spark/blob/6d32472b2a4225bd847953a679d3be2b077d212e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L1008-L1031

Comment (cloud-fan) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:204]: what's the behavior of SQL CAST? We should be consistent with that

Comment (wayneguow) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:204]: It calls toDouble directly, just like in the code above `numeric.toDouble(b)`.

image

Comment (wayneguow) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:204]: But the actual difference will be very large.

Comment (wayneguow) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:204]: Well, I agree to be consistent with Expression Cast to avoid unnecessary trouble. Update after soon.  

(But actually if we want to ensure accurate calculations, we'd better avoid using toDouble directly.)

**Comment by wayneguow** [comment]
Test failure is not related.

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala
Comment (jackierwzhang) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:924]: Thanks for implementing this (I posted the original ticket :D)!

Is there a plan to convert Date -> TimestampNTZ (Parquet [PR](https://github.com/apache/spark/pull/44368/files#diff-e067188f85ea958fe1fdfc6f83466def9ca962ebd39ca9a86aaee1ed3110c045R102)) and Int -> Decimal (Parquet [PR](https://github.com/apache/spark/pull/44803))?

Comment (wayneguow) [reply connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:924]: Yes, I will seperate another two PR.

**Review by cloud-fan — APPROVED** [review]
LGTM if CI passes

**Comment by wayneguow** [comment]
Rebase master and waiting for CI~

**Comment by cloud-fan** [comment]
thanks, merged to master

**Pull request closed without merging**
